lib/solid_flow.rb

```diff
@@ -0,0 +1,94 @@
+module SolidFlow
+  class BuildError < StandardError; end
+  class EnqueueError < StandardError; end
+
+  class Cursor
+    attr_reader :position
+
+    def initialize(position: 0)
+      @position = position
+    end
+
+    def completed?(steps)
+      position >= steps.size
+    end
+
+    def next
+      self.class.new(position: position.succ)
+    end
+  end
+
+  class CursorSerializer < ActiveJob::Serializers::ObjectSerializer
+    def serialize?(arg)
+      arg.is_a?(SolidFlow::Cursor)
+    end
+
+    def serialize(cursor)
+      super("position" => cursor.position)
+    end
+
+    def deserialize(hash)
+      SolidFlow::Cursor.new(position: hash["position"])
+    end
+  end
+
+  class Base < ActiveJob::Base
+    attr_reader :cursor, :args, :kwargs, :steps
+
+    def perform(*args, **kwargs)
+      # Remove batch arg from last step passed in on_success, not needed
+      args.shift if args.first.is_a?(SolidQueue::JobBatch)
+
+      if args.first.is_a?(SolidFlow::Cursor)
+        @cursor = args.shift
+      else
+        @cursor = Cursor.new
+      end
+
+      @args = args
+      @kwargs = kwargs
+      @steps = []
+
+      build(*args, **kwargs)
+      raise BuildError, "Workflow defined no steps" if steps.empty?
+
+      return if cursor.completed?(steps)
+
+      enqueue_step_jobs(within_batch: batch) do
+        ActiveJob.perform_all_later(step_jobs)
+      end
+    end
+
+    def enqueue_step_jobs(within_batch:, &)
+      if within_batch
+        within_batch.enqueue { SolidQueue::JobBatch.enqueue(on_success: next_job, &) }
+      else
+        SolidQueue::JobBatch.enqueue(on_success: next_job, &)
+      end
+    end
```

Comment:
Since we're within the job here, which has a batch method on it, I might just inline the argument:

```suggestion
    def enqueue_step_jobs(&)
      if batch
        batch.enqueue { SolidQueue::JobBatch.enqueue(on_success: next_job, &) }
      else
        SolidQueue::JobBatch.enqueue(on_success: next_job, &)
      end
    end
```

Reply:
Also wonder if we could get away with something like this? No idea though …it just seems like the batch nesting is something SolidQueue's batching should have an answer to.

```suggestion
    def enqueue_step_jobs(&)
      (batch || SolidQueue::JobBatch).enqueue(on_success: next_job, &)
    end
```